File: app/src/main/java/com/sjn/stamp/ui/observer/StampEditStateObserver.java
package com.sjn.stamp.ui.observer;

import com.sjn.stamp.utils.LogHelper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;


public class StampEditStateObserver {

    public enum State {
        OPEN,
        CLOSE
    }

    private static final String TAG = LogHelper.makeLogTag(StampEditStateObserver.class);

    private static StampEditStateObserver sInstance;

    State mState = State.CLOSE;

    public List<String> getSelectedStampList() {
        return mSelectedStampList;
    }

    List<String> mSelectedStampList = new ArrayList<>();

    private List<Listener> mListenerList = Collections.synchronizedList(new ArrayList<Listener>());

    public boolean isOpen() {
        return mState == State.OPEN;
    }

    public void notifyAllStampChange(String stamp) {
        LogHelper.i(TAG, "notifyAllStampChange ", mListenerList.size());
        if (mListenerList != null) {
            List<Listener> tempList = new ArrayList<>(mListenerList);
            for (Listener listener : tempList) {
                listener.onNewStampCreated(stamp);
            }
        }
    }

    public void notifySelectedStampListChange(List<String> stampList) {
        LogHelper.i(TAG, "notifySelectedStampListChange ", mListenerList.size());
        mSelectedStampList = stampList;
        if (mListenerList != null) {
            List<Listener> tempList = new ArrayList<>(mListenerList);
            for (Listener listener : tempList) {
                listener.onSelectedStampChange(mSelectedStampList);
            }
        }
    }

    public void notifyStateChange(State state) {
        LogHelper.i(TAG, "notifyStateChange ", state);
        mState = state;
        if (mListenerList != null) {
            List<Listener> tempList = new ArrayList<>(mListenerList);
            for (Listener listener : tempList) {
                listener.onStampStateChange(mState);
            }
        }
    }

    public interface Listener {
        void onSelectedStampChange(List<String> selectedStampList);

        void onNewStampCreated(String stamp);

        void onStampStateChange(State state);
    }

    public void addListener(Listener listener) {
        if (mListenerList.contains(listener)) {
            return;
        }
        mListenerList.add(listener);
    }

    public void removeListener(Listener listener) {
        if (!mListenerList.contains(listener)) {
            return;
        }
        mListenerList.remove(listener);
    }

    public static StampEditStateObserver getInstance() {
        if (sInstance == null) {
            sInstance = new StampEditStateObserver();
        }
        return sInstance;
    }

    private StampEditStateObserver() {
    }

}

File: app/src/main/java/com/sjn/stamp/ui/observer/StampEditStateObserver.kt
package com.sjn.stamp.ui.observer

import com.sjn.stamp.utils.LogHelper
import java.util.*


object StampEditStateObserver {

    private val TAG = LogHelper.makeLogTag(StampEditStateObserver::class.java)
    private var state = State.NO_EDIT

    var selectedStampList: List<String> = ArrayList()
        internal set

    private val listenerList = Collections.synchronizedList(ArrayList<Listener>())

    val isStampMode: Boolean
        get() = state == State.EDITING || state == State.STAMPING

    enum class State {
        EDITING,
        NO_EDIT,
        STAMPING
    }

    fun notifyAllStampChange(stamp: String) {
        LogHelper.i(TAG, "notifyAllStampChange ", listenerList!!.size)
        for (listener in ArrayList(listenerList)) {
            listener.onNewStampCreated(stamp)
        }
    }

    fun notifySelectedStampListChange(stampList: List<String>) {
        LogHelper.i(TAG, "notifySelectedStampListChange ", listenerList!!.size)
        selectedStampList = stampList
        for (listener in ArrayList(listenerList)) {
            listener.onSelectedStampChange(selectedStampList)
        }
    }

    fun notifyStateChange(state: State) {
        LogHelper.i(TAG, "notifyStateChange ", state)
        this.state = state
        for (listener in ArrayList(listenerList)) {
            listener.onStampStateChange(this.state)
        }
    }

    interface Listener {
        fun onSelectedStampChange(selectedStampList: List<String>)

        fun onNewStampCreated(stamp: String)

        fun onStampStateChange(state: State)
    }

    fun addListener(listener: Listener) {
        if (!listenerList.contains(listener)) {
            listenerList.add(listener)
        }
    }

    fun removeListener(listener: Listener) {
        if (listenerList.contains(listener)) {
            listenerList.remove(listener)
        }
    }

}
